---
title: "VplotR"
author: "Jacques Serizay"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
    %\VignetteIndexEntry{VplotR}
    %\VignetteEngine{knitr::rmarkdown}
    %\VignetteEncoding{UTF-8}
---

```{r, eval = TRUE, echo=FALSE, results="hide", warning=FALSE}
knitr::opts_chunk$set(
    collapse = TRUE,
    comment = "#>"
)
suppressPackageStartupMessages({
    library(GenomicRanges)
    library(ggplot2)
    library(magrittr)
    library(VplotR)
})
```

![](https://raw.githubusercontent.com/js2264/VplotR/master/man/figures/Vplots.png)

## Importing data

Paired-end .bam files are read using the `importPEBamFiles()` 
function as follows:

```{r eval = FALSE}
library(VplotR)
data(ce11_proms)
ce11_proms
fragments <- importPEBamFiles(
    'ATAC-seq-mapped-fragments.bam', 
    where = ce11_proms, 
    shift_ATAC_fragments = TRUE
)
```

*Note: to allow for a accurate background normalization, 
the `where` argument should be omitted.*

Toy datasets are available for this package: 

```{r, eval = TRUE}
data(REB1_sacCer3)
MNase_sacCer3 <- readRDS(url(
    'https://ahringerlab.com/VplotR/MNase_sacCer3_Henikoff2011.rds'
))
```

## Fragment size distribution

A preliminary control to check the distribution of fragment
sizes (regardless of their location relative to genomic loci) can be 
performed using the `getFragmentsDistribution()` function.

```{r, eval = TRUE}
df <- getFragmentsDistribution(
    MNase_sacCer3, 
    REB1_sacCer3
)
ggplot(df, aes(x = x, y = y)) + geom_line()
```

## Vplot(s)

Once data is imported, a V-plot of paired-end fragments over loci of 
interest is generated using the `plotVmat()` function:

```{r, eval = TRUE}
plotVmat(MNase_sacCer3, REB1_sacCer3)
```

The generation of multiple V-plots can be parallelized using a list of 
parameters:

```{r, eval = TRUE}
list_params <- list(
    "MNase\n@ REB1" = list(MNase_sacCer3, REB1_sacCer3), 
    "MNase\n@ random loci" = list(MNase_sacCer3, sampleGRanges(REB1_sacCer3))
)
plotVmat(
    list_params, 
    cores = 2
)
```

## Vplots normalization

Different normalization approaches are available: 

```{r, eval = TRUE}
# No normalization 
plotVmat(
    MNase_sacCer3,
    REB1_sacCer3,
    normFun = ''
)
# Library depth + number of loci of interest 
plotVmat(
    MNase_sacCer3,
    REB1_sacCer3,
    normFun = 'libdepth+nloci'
)
# Zscore
plotVmat(
    MNase_sacCer3,
    REB1_sacCer3,
    normFun = 'zscore'
)
# Quantiles
plotVmat(
    MNase_sacCer3,
    REB1_sacCer3,
    normFun = 'quantile', 
    s = 0.99
)
```

## Footprints

VplotR also implements a function to profile the footprint from MNase or 
ATAC-seq over sets of genomic loci. For instance, CTCF is known for its 
~40-bp large footprint at its binding loci. 

```{r, eval = TRUE}
data(CTCF_hg38)
ATAC_hg38 <- readRDS(url(
    'https://ahringerlab.com/VplotR/ATAC_hg38_Chen2018.rds'
))
plotFootprint(
    ATAC_hg38,
    CTCF_hg38
)
```

## Nucleosome enrichment quantification 

Finally, the `nucleosomeEnrichment()` function is useful to
quantify and compare enrichment of flanking nucleosomes at promoters 
in ATAC-seq datasets.

```{r eval = TRUE}
data(ce11_proms)
germline_proms <- ce11_proms[ce11_proms$which.tissues == "Germline"]
muscle_proms <- ce11_proms[ce11_proms$which.tissues == "Muscle"]
ATAC_ce11 <- readRDS(url(
    'https://ahringerlab.com/VplotR/ATAC_PE_fragments.rds'
))[["mixed"]]
#
nuc_enrich_germline <- nucleosomeEnrichment(
    ATAC_ce11,
    germline_proms
)
nuc_enrich_germline$fisher_test
nuc_enrich_germline$plot
nuc_enrich_muscle <- nucleosomeEnrichment(
    ATAC_ce11,
    muscle_proms
)
nuc_enrich_muscle$fisher_test
nuc_enrich_muscle$plot
```

### Session Info
```{r echo = FALSE, collapse = TRUE, eval = TRUE}
sessionInfo()
```

