---
title: "VplotR"
author: "Jacques Serizay"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
    %\VignetteIndexEntry{VplotR}
    %\VignetteEngine{knitr::rmarkdown}
    %\VignetteEncoding{UTF-8}
---

```{r, eval = TRUE, echo=FALSE, results="hide", warning=FALSE}
knitr::opts_chunk$set(
    collapse = TRUE,
    comment = "#>"
)
suppressPackageStartupMessages({
    library(GenomicRanges)
    library(ggplot2)
    library(magrittr)
    library(VplotR)
})
```

![VplotR](man/images/Vplots.png)

## Importing data

Firstly BAM files are read using the `importPEBamFiles()` function and 
loci of interest from a BED file, for instance.

```{r eval = FALSE}
library(VplotR)
granges <- rtracklayer::import('loci_of_interest.bed')
fragments <- importPEBamFiles(
    'ATAC-seq-mapped-fragments.bam', 
    where = GenomicRanges::resize(granges, width = 2000, fix = 'center'), 
    shift_ATAC_fragments = TRUE
)
```

*Note: to allow for a background normalization, 
the `where` argument should be omitted.*

Toy datasets are available for this package: 

```{r, eval = TRUE}
data(ce11_proms)
granges <- ce11_proms[ce11_proms$which.tissues == "Ubiq."]
fragments <- readRDS(url(
    'https://ahringerlab.com/VplotR/ATAC_PE_fragments.rds'
))
fragments <- fragments[["mixed"]]
```

## Fragment size distribution

 V-plots readily reveal the different patterns of fragment 
density, but a preliminary control is checking the distribution of fragment
sizes, regardless of their location relative to genomic loci. 
The getFragmentsDistribution() is useful to check this distribution.

```{r, eval = TRUE}
d <- getFragmentsDistribution(
    fragments, 
    list(
        'Ubiq.' = ce11_proms[ce11_proms$which.tissues == 'Ubiq.'],
        'Germline' = ce11_proms[ce11_proms$which.tissues == 'Germline'],
        'Muscle' = ce11_proms[ce11_proms$which.tissues == 'Muscle']
    )
)
ggplot(d, aes(x = x, y = y)) + 
    geom_line() +
    facet_wrap(~class, scales = 'free')
```

## Quick Vplot(s)

Once data is imported, a V-plot of fragments (`fragments`) 
over the loci of interest (`granges`) is generated 
using the `plotVmat()` function:

```{r}
plotVmat(fragments, granges)
```

The generation of multiple V-plots can be parallelized as follows:

```{r eval = FALSE}
list_params <- list(
    "sample_1" = list(fragments_1, granges_1), 
    "sample_2" = list(fragments_2, granges_2), 
    ..., 
    "sample_N" = list(fragments_N, granges_N)
)
plotVmat(
    list_params, 
    cores = length(list_params)
)
```

With the example datasets: 

```{r, eval = TRUE}
granges <- list(
    "Ubiq." = ce11_proms[ce11_proms$which.tissues == "Ubiq."], 
    "Germline" = ce11_proms[ce11_proms$which.tissues == "Germline"],
    "Muscle" = ce11_proms[ce11_proms$which.tissues == "Muscle"]
)
list_params <- list(
    "mixed ATAC-seq\n@ Ubiq. proms" = 
        list(fragments, granges[["Ubiq."]]), 
    "mixed ATAC-seq\n@ Germline proms" = 
        list(fragments, granges[["Germline"]]), 
    "mixed ATAC-seq\n@ Muscle proms" = 
        list(fragments, granges[["Muscle"]])
)
plotVmat(
    list_params, 
    cores = 1
)
```

## Vplots normalization

Different normalization approaches are available: 

```{r, eval = TRUE}
granges <- ce11_proms[ce11_proms$which.tissues == "Ubiq."]
# No normalization 
plotVmat(
    fragments,
    granges,
    normFun = ''
)
# Library depth + number of loci of interest 
plotVmat(
    fragments,
    granges,
    normFun = 'libdepth+nloci'
)
# Zscore
plotVmat(
    fragments,
    granges,
    normFun = 'zscore'
)
# Quantiles
plotVmat(
    fragments,
    granges,
    normFun = 'quantile', 
    s = 0.99
)
```

## Nucleosome enrichment quantification 

Finally, the `nucleosomeEnrichment()` function is useful to statistically 
quantify and compare nucleosome enrichment (e.g. flanking nucleosome 
at promoters) from ATAC-seq datasets. To do so:

```{r eval = TRUE}
data(ce11_proms)
granges <- ce11_proms[ce11_proms$which.tissues == "Ubiq."]
fragments <- readRDS(url(
    'https://ahringerlab.com/VplotR/ATAC_PE_fragments.rds'
))
fragments <- fragments[["mixed"]]
#
set.seed(42) 
nuc_enrich <- nucleosomeEnrichment(
    fragments,
    ce11_proms[ce11_proms$which.tissues == 'Ubiq.']
)
nuc_enrich$fisher_test
nuc_enrich$plot
```

## Footprints

TBC

### Session Info
```{r echo = FALSE, collapse = TRUE, eval = TRUE}
sessionInfo()
```

